src/jit/codegenarm64.cpp

```diff
@@ -2664,92 +2664,118 @@ void CodeGen::genLockedInstructions(GenTreeOp* treeNode)
     regNumber dataReg   = data->gtRegNum;
     regNumber addrReg   = addr->gtRegNum;
 
-    regNumber exResultReg  = treeNode->ExtractTempReg(RBM_ALLINT);
-    regNumber storeDataReg = (treeNode->OperGet() == GT_XCHG) ? dataReg : treeNode->ExtractTempReg(RBM_ALLINT);
-    regNumber loadReg      = (targetReg != REG_NA) ? targetReg : storeDataReg;
+    genConsumeAddress(addr);
+    genConsumeRegs(data);
 
-    // Check allocator assumptions
-    //
-    // The register allocator should have extended the lifetimes of all input and internal registers so that
-    // none interfere with the target.
-    noway_assert(addrReg != targetReg);
+    emitAttr dataSize = emitActualTypeSize(data);
 
-    noway_assert(addrReg != loadReg);
-    noway_assert(dataReg != loadReg);
+    if (compiler->compSupports(InstructionSet_Atomics))
+    {
+        assert(!data->isContainedIntOrIImmed());
 
-    noway_assert(addrReg != storeDataReg);
-    noway_assert((treeNode->OperGet() == GT_XCHG) || (addrReg != dataReg));
+        switch (treeNode->gtOper)
+        {
+            case GT_XCHG:
+                getEmitter()->emitIns_R_R_R(INS_swpal, dataSize, dataReg, targetReg, addrReg);
+                break;
+            case GT_XADD:
+                if ((targetReg == REG_NA) || (targetReg == REG_ZR))
```

Comment:
Can `targetReg` really be `REG_NA`?!

Reply:
Looks like `staddl` is the arm64 equivalent of x64's `lock add`. If you want to support this you need to generate `GT_LOCKADD` in lowering like it is done on x64. See #18267

Reply:
If the result is not used, it certainly could be REG_NA/REG_ZR.  I doubt lsra and/or lowering will generate this case yet.

Reply:
Before the rebase this used to include GT_LOCKADD.   I will look at #18267